File: dijkstra/dijkstra.py
from queue import PriorityQueue

class node:
	name = ""
	cost = 9999
	seen = False
	back = None

	def __init__(self, name):
		self.connected = []
		self.name = name
		pass

	def __gt__(self, other):
		return self.cost > other.cost

	def __str__(self):
		return f"Node(name={self.name}, cost={self.cost})"

with open("edgelist.txt") as file:
	for line in file:
		a, b, weight = [ int(v) for v in line.strip().split(",") ]

def walk(init, target):
	queue = PriorityQueue() # Маленькие первыми
	queue.put(init)

	init.cost = 0

	# Перебираем соседей
	# обновляем стоимость
	# заполняем очередь
	while not queue.empty():
		loc = queue.get()
		loc.seen = True

		print(f"In {loc}")

		for dst in loc.connected:
			if dst.seen:
				continue

			linkage_key = f"{loc.name},{dst.name}"
			cost_from_here = loc.cost + weights[linkage_key]

			if cost_from_here < dst.cost:
				dst.cost = cost_from_here
				dst.back = loc

			print(f"Establish that {dst}")

			queue.put(dst)

	# Пробегаемся от цели к началу
	hist = []
	loc = target

	while True:
		hist.append(loc.name)
		if loc.back:
			loc = loc.back
		else:
			break

	hist.reverse()

	print(f"Path from {init.name} to {target.name} costs {target.cost}: {hist}")

weights = {
	"1,2": 10,
	"1,3": 40,
	"2,3": 10
}

a = node("1")
b = node("2")
c = node("3")

a.connected.append(b)
a.connected.append(c)
b.connected.append(c)

walk(a, c)

File: dijkstra/dijkstra.cpp

#include <cassert>
#include <cstdio>
#include <vector>
#include <queue>


//
// Вершина
//
class node {
public:

	// Ребро
	class linkage {
	public:
		node* dst;
		int weight;

		// Конструктор
		linkage(node* dst, int weight) : dst(dst), weight(weight) {
		}
	};

	std::vector<linkage> connected;
	bool seen = false;
	int cost = 9999;
	int name = 0;

	node* back = nullptr;

	// Конструктор
	node(int name) : name(name) {
	}
};

//
// Граф
//
class graph {
public:
	std::vector<node*> nodes;

	node* node_by_name(int name) {
		for (node* it : nodes) {
			if ( it->name == name )
				return it;
		}
		return nullptr;
	}

	node* add_node(int name) {
		node* result = new node(name);
		nodes.push_back(result);
		return result;
	}
};

class node_prio_queue_entry {
public:
	node* node_ptr;

	// Конструктор
	node_prio_queue_entry(node* node_ptr) : node_ptr(node_ptr) {
	}

	// Для упорядочивания
	bool operator<(const node_prio_queue_entry& other) const {
		return node_ptr->cost > other.node_ptr->cost;
	}
};

//
// Дейкстра
//
void walk(node* init, node* target) {
	std::priority_queue<node_prio_queue_entry> queue;

	init->cost = 0;

	queue.push( node_prio_queue_entry(init) );

	// Пробежка
	while (!queue.empty()) {
		node_prio_queue_entry const& e = queue.top();
		node* loc = e.node_ptr;
		queue.pop();

		loc->seen = true;

		printf("In node(%d) cost %d\n", loc->name, loc->cost);

		for (node::linkage link : loc->connected) {
			node* dst = link.dst;

			if (dst->seen)
				continue;

			int cost_from_here = loc->cost + link.weight;

			if (cost_from_here < dst->cost) {
				dst->cost = cost_from_here;
				dst->back = loc;
			}

			printf("Establish that node(%d) cost %d\n", dst->name, dst->cost);

			queue.push( node_prio_queue_entry(link.dst) );
		}
	}

	printf("Path from %d to %d costs %d\n", init->name, target->name, target->cost);

	// Пробежка от цели к началу
	std::vector<node*> history;
	node* loc = target;

	while (1) {
		history.push_back(loc);
		loc = loc->back;
		if (!loc) break;
	}

	for (auto it = history.rbegin(); it != history.rend(); it++) {
		printf("%d ", (*it)->name);
	}

	printf("End\n");
}

int main() {
	FILE* fd = fopen("dijkstra/edgelist.txt", "r");
	graph g;

	while (!feof(fd)) {
		int a, b, w;

		int rc = fscanf(fd, "%d,%d,%d\n", &a, &b, &w);

		assert(rc == 3);

		node* a_ = g.node_by_name(a);
		node* b_ = g.node_by_name(b);

		if (!a_)
			a_ = g.add_node(a);

		if (!b_)
			b_ = g.add_node(b);

		a_->connected.push_back( node::linkage(b_, w) );
	}

	walk( g.node_by_name(1), g.node_by_name(9) );
}
